File: packages/algorithm/Leetcode/hot100/数据结构/二叉树/二叉树的直径.ts
import {TreeNode} from "./types";


// 直径 = 左最大深度 + 右最大深度
function diameterOfBinaryTree(root: TreeNode | null): number {
    let res = 0;

    const dfs = (node: TreeNode | null): number => {
        if (node === null) {
            return 0
        }

        const left = dfs(node.left);
        const right = dfs(node.right);

        // 左右最大深度相加
        res = Math.max(res, left + right);

        return Math.max(left, right) + 1;
    }

    dfs(root);

    return res;
};

File: packages/program-design/alog/数据结构/二叉树/py/二叉树的直径.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        ans = 0
        def dfs(node: Optional[TreeNode]):
            nonlocal ans
            if not node:
                return 0

            cnt_left = dfs(node.left)
            cnt_right = dfs(node.right)
            ans = max(ans, cnt_right + cnt_left)

            return 1 + max(cnt_left, cnt_right)

        dfs(root)

        return ans